torch/csrc/distributed/c10d/Functional.cpp

```diff
@@ -0,0 +1,280 @@
+    {"max", c10d::ReduceOp(c10d::ReduceOp::RedOpType::MAX)},
+    {"band", c10d::ReduceOp(c10d::ReduceOp::RedOpType::BAND)},
+    {"bor", c10d::ReduceOp(c10d::ReduceOp::RedOpType::BOR)},
+    {"bxor", c10d::ReduceOp(c10d::ReduceOp::RedOpType::BXOR)},
+    // {"premul_sum", c10d::ReduceOp(c10d::ReduceOp::RedOpType::PREMUL_SUM)},
+    {"unused", c10d::ReduceOp(c10d::ReduceOp::RedOpType::UNUSED)}};
+
+at::Tensor all_reduce_(
+    at::Tensor input,
+    const std::string& reduceOp,
+    const std::string& tag,
+    const std::vector<int64_t> ranks,
+    const int64_t groupSize) {
+  c10d::AllreduceOptions opts;
+  auto reduceOpIt = strToReduceOp.find(reduceOp);
+  if (reduceOpIt == strToReduceOp.end()) {
+    LOG(FATAL) << "Unrecognized reduceOp: " << reduceOp;
+  }
+  opts.reduceOp = reduceOpIt->second;
+
+  std::vector<at::Tensor> inputs{input};
+  auto pg = c10d::ProcessGroup::resolveFromName(tag);
+  if (groupSize != pg->getSize()) {
+    LOG(FATAL) << "Mismatch between group size argument and world size"
+                  "of process group ("
+               << tag << ")";
+  }
+  auto work = pg->allreduce(inputs, opts);
+  pendingWork[input.data_ptr()] = work;
+  return input;
+}
+
+at::Tensor all_reduce(
+    const at::Tensor& input,
+    const std::string& reduceOp,
+    const std::string& tag,
+    const std::vector<int64_t> ranks,
+    const int64_t groupSize) {
+  auto output = input.clone();
+  return all_reduce_(output, reduceOp, tag, ranks, groupSize);
+}
+
+std::vector<at::Tensor> all_reduce_coalesced_(
+    std::vector<at::Tensor> inputs,
+    const std::string& reduceOp,
+    const std::string& tag,
+    const std::vector<int64_t> ranks,
+    const int64_t groupSize) {
+  c10d::AllreduceCoalescedOptions opts;
+  auto reduceOpIt = strToReduceOp.find(reduceOp);
+  if (reduceOpIt == strToReduceOp.end()) {
+    LOG(FATAL) << "Unrecognized reduceOp: " << reduceOp;
+  }
+  opts.reduceOp = reduceOpIt->second;
+
+  auto pg = c10d::ProcessGroup::resolveFromName(tag);
+  if (groupSize != pg->getSize()) {
+    LOG(FATAL) << "Mismatch between group size argument and world size"
+                  "of process group ("
+               << tag << ")";
+  }
+  auto work = pg->allreduce_coalesced(inputs, opts);
+  for (const auto& tensor : inputs) {
+    // c10d::Work::wait is idempotent
+    pendingWork[tensor.data_ptr()] = work;
+  }
+  return inputs;
+}
+
+std::vector<at::Tensor> all_reduce_coalesced(
+    const std::vector<at::Tensor>& inputs,
+    const std::string& reduceOp,
+    const std::string& tag,
+    const std::vector<int64_t> ranks,
+    const int64_t groupSize) {
+  std::vector<at::Tensor> outputs;
+  for (const auto& tensor : inputs) {
+    outputs.push_back(tensor.clone());
+  }
+  return all_reduce_coalesced_(outputs, reduceOp, tag, ranks, groupSize);
+}
+
+at::Tensor allocateAllGatherOutput(const at::Tensor& input, int64_t groupSize) {
+  auto outputSize = input.sizes().vec();
+  outputSize[0] *= groupSize;
```

Comment:
is AllGather always expanding on the first dim ?

Reply:
Yes. This is the behavior for the current API.